beacon_node/operation_pool/src/lib.rs

```diff
@@ -50,6 +67,116 @@ impl<T: EthSpec> OperationPool<T> {
+        contribution: SyncCommitteeContribution<T>,
+        fork: &Fork,
+        genesis_validators_root: Hash256,
+        spec: &ChainSpec,
+    ) -> Result<(), OpPoolError> {
+        let aggregate_id = SyncAggregateId::from_data::<T>(
+            contribution.slot,
+            contribution.beacon_block_root,
+            fork,
+            genesis_validators_root,
+            spec,
+        );
+
+        let mut contributions = self.sync_contributions.write();
+
+        match contributions.entry(aggregate_id) {
+            Entry::Vacant(entry) => {
+                // If no contributions or aggregate exist for these keys, insert both.
+                let contributions = vec![contribution];
+                let aggregate = SyncAggregate::from_contributions(contributions.as_slice())?;
+                entry.insert((contributions, aggregate));
+            }
+            Entry::Occupied(mut entry) => {
+                // If both a contribution and aggregate exist for this key, check whether the new or
+                // old contribution has more aggregation bits set. If the new one does, add it to the
+                // aggregate and insert both the aggregate and the contribution.
+                let existing_contributions = entry.get_mut();
+                match existing_contributions
+                    .0
+                    .iter()
+                    .position(|existing_contribution| {
+                        existing_contribution.subcommittee_index == contribution.subcommittee_index
+                    }) {
+                    Some(position) => {
+                        // Only need to recalculate if the new contribution has more bits set.
+                        if existing_contributions.0[position].aggregation_bits.len()
+                            < contribution.aggregation_bits.len()
+                        {
+                            existing_contributions.0[position] = contribution;
+                            let aggregate = SyncAggregate::from_contributions(
+                                existing_contributions.0.as_slice(),
+                            )?;
+                            existing_contributions.1 = aggregate;
+                        }
+                    }
+                    None => {
+                        // If there has been no previous sync contribution for this subcommittee index,
+                        // add it and recalculate the aggregate.
+                        existing_contributions.0.push(contribution);
+                        let aggregate =
+                            SyncAggregate::from_contributions(existing_contributions.0.as_slice())?;
+                        existing_contributions.1 = aggregate;
+                    }
+                }
+            }
+        };
+        Ok(())
+    }
+
+    /// Get the a sync aggregate for inclusion in a block.
+    pub fn get_sync_aggregate(
+        &self,
+        state: &BeaconState<T>,
+        block_root: Hash256,
```

Comment:
Are you using the `block_root` argument in the tests? I refactored this function in the VC branch so that it just pulls the block root from the state: https://github.com/michaelsproul/lighthouse/commit/d1ac8b91899e31c32121c62720c7941c135bb119

In alpha.6 at least, the aggregate is for the slot before.

Reply:
This is good with me, won't impact tests. But wouldn't `state.slot()` here already be for the prior slot? Because this is called before block production on the current slot?

Reply:
Ah nevermind, I now see that `complete_state_advance` happens first